PhysicsTools/PatAlgos/plugins/PackedCandidateMuonSelectorProducer.cc

```diff
@@ -0,0 +1,122 @@
+      for (const auto& sel : muonSelectors_) {
+        produces<BoolMap>(sel);
+      }
+      for (const auto& sel : muonIDs_) {
+        muonIDMap_[sel].reset(new StringCutObjectSelector<pat::Muon>("passed('" + sel + "')"));
+        produces<BoolMap>(sel);
+      }
+    }
+    ~PackedCandidateMuonSelectorProducer() override{};
+
+    void produce(edm::Event&, const edm::EventSetup&) override;
+
+    static void fillDescriptions(edm::ConfigurationDescriptions&);
+
+  private:
+    edm::EDGetTokenT<pat::MuonCollection> muonToken_;
+    edm::EDGetTokenT<edm::View<pat::PackedCandidate> > candidateToken_;
+    std::vector<std::string> muonSelectors_;
+    std::vector<std::string> muonIDs_;
+    std::map<std::string, std::unique_ptr<StringCutObjectSelector<pat::Muon> > > muonIDMap_;
+    double maxDeltaR_, maxDeltaRelPt_;
+  };
+
+}  // namespace pat
+
+void pat::PackedCandidateMuonSelectorProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  edm::Handle<pat::MuonCollection> muons;
+  edm::Handle<edm::View<pat::PackedCandidate> > candidates;
+
+  iEvent.getByToken(muonToken_, muons);
+  iEvent.getByToken(candidateToken_, candidates);
+
+  const auto& nCand = candidates->size();
+
+  std::map<std::string, std::vector<bool> > muonSelMap;
+  for (const auto& sel : muonSelectors_) {
+    muonSelMap[sel] = std::vector<bool>(nCand, false);
+  }
+  for (const auto& sel : muonIDs_) {
+    muonSelMap[sel] = std::vector<bool>(nCand, false);
+  }
+
+  // loop over muons
+  for (const auto& muon : *muons) {
+    // ignore muons without high purity inner track
+    if (muon.innerTrack().isNull() || !muon.innerTrack()->quality(reco::TrackBase::qualityByName("highPurity")))
+      continue;
+
+    // find candidate associated to muon
+    const auto& muonTrack = *muon.innerTrack();
+    for (size_t i = 0; i < nCand; i++) {
+      const auto& cand = candidates->refAt(i);
+      // ignore neutral candidates or without track
+      if (cand->charge() == 0 || !cand->hasTrackDetails())
+        continue;
+
+      // check if candidate and muon are compatible
+      const auto& candTrack = cand->pseudoTrack();
+      if (muonTrack.charge() == candTrack.charge() && muonTrack.numberOfValidHits() == candTrack.numberOfValidHits() &&
+          reco::deltaR(muonTrack.eta(), muonTrack.phi(), candTrack.eta(), candTrack.phi()) < maxDeltaR_ &&
+          std::abs((muonTrack.pt() - candTrack.pt()) / muonTrack.pt()) < maxDeltaRelPt_) {
+        // fill muon selector map
+        for (const auto& sel : muonSelectors_) {
+          muonSelMap[sel][i] = muon.muonID(sel);
+        }
+        for (const auto& sel : muonIDs_) {
+          muonSelMap[sel][i] = (*muonIDMap_.at(sel))(muon);
+        }
+        break;
+      }
+    }
+  }
+
+  // fill the value maps
+  for (const auto& s : muonSelMap) {
+    std::unique_ptr<BoolMap> valueMap(new BoolMap());
+    BoolMap::Filler filler(*valueMap);
```

Comment:
are the ValueMaps compressing well enough?
IIUC, the ratio is perhaps 100:1 of false:true.
Saving a RefVector for passing candidates will be much smaller.
It sounds like the more likely case is to read from these selected muon-like packed cands, rather than reading the full packed cand collection to check/reject the case with muons.
A RefVector will be also easier to use in that case.

Reply:
Changed the output to a RefVector as suggested. Running over ~1000 events from the HIDoubleMuon PD of 2018 PbPb data, the size of the output went in the case of the packedPFCandidates from an average compressed size of ~160 bytes/event for valueMaps down to 9 bytes/event for the PackedCandidateRefVectors, and for the case of lostTracks from ~38 bytes/event for valueMaps down to 6 bytes/event for the PackedCandidateRefVectors.

Reply:
please copy/add the summary about the event sizes to the PR description.
Thank you.